Fix container restart backoff

wzshiming opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
/sig node

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #99697

#### Special notes for your reviewer:

<details>
<summary>Pod yaml</summary>

``` yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    test: liveness
  name: liveness-exec
  namespace: default
spec:
  containers:
    - name: c1
      image: alpine
      imagePullPolicy: IfNotPresent
      args:
        - /bin/sh
        - -c
        - sleep 1800
      startupProbe:
        exec:
          command:
            - "false"
        initialDelaySeconds: 1
        periodSeconds: 1
        failureThreshold: 1
```

</details>

<details>
<summary>Before</summary>

``` 
NAME            READY   STATUS    RESTARTS   AGE
liveness-exec   0/1     Running            0          12s
liveness-exec   0/1     Running            1          34s
liveness-exec   0/1     Running            2          66s
liveness-exec   0/1     Running            3          99s
liveness-exec   0/1     Running            4          2m12s
liveness-exec   0/1     CrashLoopBackOff   4          2m43s
liveness-exec   0/1     Running            5          3m28s
liveness-exec   0/1     CrashLoopBackOff   5          3m59s
liveness-exec   0/1     Running            6          5m32s
liveness-exec   0/1     CrashLoopBackOff   6          6m4s
liveness-exec   0/1     Running            7          8m48s
liveness-exec   0/1     CrashLoopBackOff   7          9m20s
liveness-exec   0/1     Running            8          14m
liveness-exec   0/1     Running            9          15m
liveness-exec   0/1     CrashLoopBackOff   9          15m
liveness-exec   0/1     Running            10         20m
liveness-exec   0/1     Running            11         21m
liveness-exec   0/1     CrashLoopBackOff   11         21m
liveness-exec   0/1     Running            12         26m
liveness-exec   0/1     Running            13         27m
liveness-exec   0/1     CrashLoopBackOff   13         27m
liveness-exec   0/1     Running            14         32m
liveness-exec   0/1     Running            15         33m
liveness-exec   0/1     CrashLoopBackOff   15         33m
liveness-exec   0/1     Running            16         39m
liveness-exec   0/1     Running            17         39m
liveness-exec   0/1     CrashLoopBackOff   17         40m
liveness-exec   0/1     Running            18         45m
liveness-exec   0/1     Running            19         45m
liveness-exec   0/1     CrashLoopBackOff   19         46m
```
</details>

<details>
<summary>After</summary>


``` 
NAME            READY   STATUS    RESTARTS   AGE
liveness-exec   0/1     Running            1          33s
liveness-exec   0/1     Running            2          66s
liveness-exec   0/1     Running            3          98s
liveness-exec   0/1     Running            4          2m11s
liveness-exec   0/1     CrashLoopBackOff   4          2m42s
liveness-exec   0/1     Running            5          2m56s
liveness-exec   0/1     CrashLoopBackOff   5          3m27s
liveness-exec   0/1     Running            6          4m23s
liveness-exec   0/1     CrashLoopBackOff   6          4m54s
liveness-exec   0/1     Running            7          7m8s
liveness-exec   0/1     CrashLoopBackOff   7          7m39s
liveness-exec   0/1     Running            8          12m
liveness-exec   0/1     CrashLoopBackOff   8          12m
liveness-exec   0/1     Running            9          17m
liveness-exec   0/1     CrashLoopBackOff   9          17m
liveness-exec   0/1     Running            10         22m
liveness-exec   0/1     CrashLoopBackOff   10         23m
liveness-exec   0/1     Running            11         27m
liveness-exec   0/1     CrashLoopBackOff   11         28m
```

</details>


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix container restart backoff
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

wzshiming committed:
Add test for BackOff

wzshiming commented:
/cc @matthyx

matthyx commented:
/triage accepted
/priority important-longterm

matthyx commented:
@wzshiming thanks for the PR... could you please elaborate with `time.Now()` is more suitable than `cStatus.FinishedAt` in this case? What's the rationale behind this choice?

wzshiming committed:
Fix container restart backoff

wzshiming commented:
> @wzshiming thanks for the PR... could you please elaborate why `time.Now()` is more suitable than `cStatus.FinishedAt` in this case? What's the rationale behind this choice?

After the actual test, there was a problem when restarting 3 times, and it could not be restarted anymore. Checking the date found that the test always output `"Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(beb742a7-ca09-4744-8bd5-11591edfcf8d)"`

I think it stays at the 40s stage and cannot enter the next stage because cStatus.FinishedAt is fixed

matthyx commented:
> I think it stays at the 40s stage and cannot enter the next stage because cStatus.FinishedAt is fixed

So maybe the bug is that `cStatus.FinishedAt` isn't reset?

wzshiming commented:
1. I added test to `IsInBackOffSinceUpdate` and `IsInBackOffSince`, and found that the behavior of `IsInBackOffSince` is a bit strange. And `IsInBackOffSinceUpdate` meets the expected backoff, I first changed `IsInBackOffSince` to `IsInBackOffSinceUpdate`  https://github.com/kubernetes/kubernetes/commit/d20b60018d1cb76ada371ea47611bd720074b4a4#diff-ab6201fe6ab966b661c5a6ea1afb7c27ba57ae1a3b6c558d9d820813c91e2c64R269-R339
2. Then test https://github.com/kubernetes/kubernetes/pull/103257#issuecomment-869631987
3. I checked the ones that currently use `IsInBackOffSinceUpdate` and they all use `time.Now()`. I changed it to `time.Now()`.
4. After testing, the backoff behavior was as expected.

matthyx commented:
> So maybe the bug is that `cStatus.FinishedAt` isn't reset?

Sorry @wzshiming I still think this is the root of the issue... so not using it actually helps here but doesn't solve the underlying issue IMHO

wzshiming commented:
Maybe not, If the time passed to `IsInBackOffSinceUpdate` does not change, and `Next` is not called, then the result returned by `IsInBackOffSinceUpdate` will not change. `FinishedAt` of the container remains unchanged until the next container. so it has been in the same time window.



Below is I added debug log and watch pod event. `FinishedAt` has been reset.

``` log
liveness-exec   0/1     Running             0          2s
liveness-exec   0/1     Running             1          35s
liveness-exec   0/1     Running             2          67s
liveness-exec   0/1     CrashLoopBackOff    2          98s
liveness-exec   0/1     Running             3          99s
liveness-exec   0/1     CrashLoopBackOff    3          2m11s
liveness-exec   0/1     Running             4          2m12s
liveness-exec   0/1     CrashLoopBackOff    4          2m43s
```

``` log
Jun 28 22:24:32 zsm-test-1 kubelet[69061]: I0628 22:24:32.508057   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:24:00.207572243 +0800 CST" RestartCount=0
Jun 28 22:25:04 zsm-test-1 kubelet[69061]: I0628 22:25:04.829863   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:24:32.499582066 +0800 CST" RestartCount=1
Jun 28 22:25:04 zsm-test-1 kubelet[69061]: I0628 22:25:04.830025   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 10s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:25:04 zsm-test-1 kubelet[69061]: I0628 22:25:04.951014   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:25:04.820234677 +0800 CST" RestartCount=2
Jun 28 22:25:37 zsm-test-1 kubelet[69061]: I0628 22:25:37.076941   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:25:04.820234677 +0800 CST" RestartCount=2
Jun 28 22:25:37 zsm-test-1 kubelet[69061]: I0628 22:25:37.077139   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 20s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:25:37 zsm-test-1 kubelet[69061]: I0628 22:25:37.186924   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:25:37.070102815 +0800 CST" RestartCount=3
Jun 28 22:26:09 zsm-test-1 kubelet[69061]: I0628 22:26:09.295399   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:25:37.070102815 +0800 CST" RestartCount=3
Jun 28 22:26:09 zsm-test-1 kubelet[69061]: I0628 22:26:09.295589   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:09 zsm-test-1 kubelet[69061]: I0628 22:26:09.432916   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:09 zsm-test-1 kubelet[69061]: I0628 22:26:09.433131   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:10 zsm-test-1 kubelet[69061]: I0628 22:26:10.457750   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:10 zsm-test-1 kubelet[69061]: I0628 22:26:10.457939   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:23 zsm-test-1 kubelet[69061]: I0628 22:26:23.946299   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:23 zsm-test-1 kubelet[69061]: I0628 22:26:23.946562   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:34 zsm-test-1 kubelet[69061]: I0628 22:26:34.947107   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:34 zsm-test-1 kubelet[69061]: I0628 22:26:34.947650   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:48 zsm-test-1 kubelet[69061]: I0628 22:26:48.946538   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:48 zsm-test-1 kubelet[69061]: I0628 22:26:48.947070   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:27:01 zsm-test-1 kubelet[69061]: I0628 22:27:01.945863   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:27:01 zsm-test-1 kubelet[69061]: I0628 22:27:01.946132   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:27:12 zsm-test-1 kubelet[69061]: I0628 22:27:12.946387   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:27:12 zsm-test-1 kubelet[69061]: I0628 22:27:12.946652   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:27:24 zsm-test-1 kubelet[69061]: I0628 22:27:24.946833   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:27:24 zsm-test-1 kubelet[69061]: I0628 22:27:24.947091   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:27:35 zsm-test-1 kubelet[69061]: I0628 22:27:35.945933   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:27:35 zsm-test-1 kubelet[69061]: I0628 22:27:35.946231   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
```

matthyx commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -932,6 +932,8 @@ func (m *kubeGenericRuntimeManager) doBackOff(pod *v1.Pod, container *v1.Contain
 
+	// Use the finished time of the latest exited container as the start point to calculate whether to do back-off.
+	ts := cStatus.FinishedAt
 	backOff.Next(key, ts)
```
for consistency, since `backoff.Next()` also checks expiration you should use `now` I guess

matthyx commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -919,11 +919,11 @@ func (m *kubeGenericRuntimeManager) doBackOff(pod *v1.Pod, container *v1.Contain
 	key := getStableKey(pod, container)
-	if backOff.IsInBackOffSince(key, ts) {
+	now := time.Now()
+	if backOff.IsInBackOffSinceUpdate(key, now) {
```
ok, changing to `IsInBackOffSinceUpdate` makes it working

derekwaynecarr replied:
I do not understand the reasoning for no longer looking at the finished timestamp of the last exited container versus the present timestamp, how did that change fix the core issue?  can you elaborate?

matthyx commented:
> Maybe not, If the time passed to `IsInBackOffSinceUpdate` does not change, and `Next` is not called, then the result returned by `IsInBackOffSinceUpdate` will not change. `FinishedAt` of the container remains unchanged until the next container. so it has been in the same time window.

Yes, got it now... sorry for the delay in understanding. I think indeed it's right to check with now, pls check my inline comments.

wzshiming commented:
@matthyx  Thanks, updated.

matthyx commented:
I think it's correct now... let's see what the CI says...
/lgtm

wzshiming commented:
/assign @derekwaynecarr @sttts

wzshiming commented:
/assign @liggitt

wzshiming commented:
/assign @mrunalp

liggitt commented:
Is the issue specific to failing startup probes? Are existing crash-loop scenarios where the container actually finishes and crashes working properly? If so, can we scope the fix more narrowly so that it doesn't change behavior for existing scenarios that are working?

wzshiming commented:
> Is the issue specific to failing startup probes?

All kinds of fail have this issue.

liggitt commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -919,11 +919,11 @@ func (m *kubeGenericRuntimeManager) doBackOff(pod *v1.Pod, container *v1.Contain
 	// backOff requires a unique key to identify the container.
 	key := getStableKey(pod, container)
-	if backOff.IsInBackOffSince(key, ts) {
+	now := time.Now()
```
this needs a test demonstrating the issue this is fixing

liggitt commented:
/hold

explicitly holding since it is not clear this is correct, and there is no test demonstrating the issue and the fix

wzshiming commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

wzshiming commented:
@liggitt 
I added the e2e test to cover the case of container restart backoff

``` log
Jul  9 20:39:41.075: INFO: container restart backoff: count:1 interval:2s startdAt:2021-07-09 20:39:30 +0800 CST
Jul  9 20:39:45.101: INFO: container restart backoff: count:2 interval:10s startdAt:2021-07-09 20:39:40 +0800 CST
Jul  9 20:40:17.238: INFO: container restart backoff: count:3 interval:4s startdAt:2021-07-09 20:39:44 +0800 CST
Jul  9 20:41:01.443: INFO: container restart backoff: count:4 interval:32s startdAt:2021-07-09 20:40:16 +0800 CST
Jul  9 20:42:30.791: INFO: container restart backoff: count:5 interval:45s startdAt:2021-07-09 20:41:01 +0800 CST
Jul  9 20:45:21.439: INFO: container restart backoff: count:6 interval:89s startdAt:2021-07-09 20:42:30 +0800 CST
Jul  9 20:50:25.635: INFO: container restart backoff: count:7 interval:171s startdAt:2021-07-09 20:45:21 +0800 CST
Jul  9 20:55:33.953: INFO: container restart backoff: count:8 interval:304s startdAt:2021-07-09 20:50:25 +0800 CST
Jul  9 21:00:30.115: INFO: container restart backoff: count:9 interval:307s startdAt:2021-07-09 20:55:32 +0800 CST
```

wzshiming commented:
/test pull-kubernetes-integration

wzshiming commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

wzshiming commented:
@liggitt Hi. could you have time to take a look

matthyx commented:
Still LGTM, maybe we can merge this after code thaw @mrunalp ?
/lgtm

derekwaynecarr commented:
I want to review this more closely.  Is it possible to split this PR into a test case that identifies the problem (and consistently fails) separate from the solution?

derekwaynecarr commented:
/hold

derekwaynecarr commented on test/e2e_node/container_restart_backoff_test.go:
```diff
@@ -0,0 +1,124 @@
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = SIGDescribe("ContainerRestartBackoff [Slow]", func() {
```
a separate pr with just this test and a skip on it is helpful that shows how back-off is consistently (or inconsistently) broken.  we have also made some related refactors in kubelet at end of this release where it would be good to accumulate signal on this test in isolation of a fix multiple times.

wzshiming replied:
OK, I sent a new PR https://github.com/kubernetes/kubernetes/pull/104140

liggitt commented on test/e2e_node/container_restart_backoff_test.go:
```diff
@@ -0,0 +1,124 @@
+				framework.Logf("container restart backoff: count:%d interval:%ds startdAt:%s", count, gotIntervalSecond, terminated.StartedAt)
+
+				// Theoretically, the maximum tolerance is backOffPeriod, but in order to avoid the flakes from increasing the tolerance
+				tolerance := int(float64(backOffPeriod) * 1.5)
```
even a 50% tolerance on a 10 second interval still seems like it could be flaky if the kubelet reconcile loop is involved

derekwaynecarr replied:
agree with @liggitt that this could be flaky when kubelet is running large numbers of concurrent pods.  if i read the original reported issue, it looked like there were periods between restarts with no backoff, is it better to just test that?  from a cluster user perspective, we make now outward commitments on retry and backoff periods generally, its more to preserve the health of the node and node agents than for the end-user benefit.

derekwaynecarr commented on staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:
```diff
@@ -193,3 +194,75 @@ func TestIsInBackOffSinceUpdate(t *testing.T) {
 	}
 }
+
+func TestIsInBackOffSinceUpdateTick(t *testing.T) {
```
this test seems distinct from the kuberuntime change, and may be useful providing in its own PR.

i am still confused on the usage of the now timestamp though versus last observed exit time.

wzshiming replied:
I sent a new PR https://github.com/kubernetes/kubernetes/pull/104139

I have the same confusion about `IsInBackOffSince`. Fortunately, there is no other place to use it except here. `IsInBackOffSinceUpdate` is used in the same way in other places. After that, we should use a clearer `IsInBackOffSinceUpdate`

SergeyKanzhelev commented:
/assign 
/assign @ehashman

ehashman commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -945,11 +945,11 @@ func (m *kubeGenericRuntimeManager) doBackOff(pod *v1.Pod, container *v1.Contain
 	// backOff requires a unique key to identify the container.
 	key := getStableKey(pod, container)
-	if backOff.IsInBackOffSince(key, ts) {
+	now := time.Now()
+	if backOff.IsInBackOffSinceUpdate(key, now) {
```
This doesn't seem right to me. We reset the time every run through this function, so won't that prevent an exponential back off? According to the output in the first comment, the restart interval is linear/fixed now, which I believe is a change in behaviour.

The root issue seems to be that there's some sort of race described in #99697 resulting in two restarts. I'm not sure how this change addresses that.

/hold

matthyx replied:
If you compare `backOff.IsInBackOffSince(key, ts)`
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/client-go/util/flowcontrol/backoff.go#L92-L103
with `backOff.IsInBackOffSinceUpdate(key, now)`
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/client-go/util/flowcontrol/backoff.go#L106-L117

You can see the only difference is on the condition `hasExpired`, so does it really changes the exponential backoff?
This might eventually come from the change that I requested when calling `backOff.Next(key, now)`.

I think we should be asking ourselves why there is a bug on the condition described above... also this func in only called from here which would make it easier to fix without side effects.

Also this calls for more investigation:

> Maybe not, If the time passed to IsInBackOffSinceUpdate does not change, and Next is not called, then the result returned by IsInBackOffSinceUpdate will not change. FinishedAt of the container remains unchanged until the next container. so it has been in the same time window.

Which is what we see happening...

ehashman commented:
/lgtm cancel

wzshiming commented:
/retest

wzshiming commented:
https://github.com/kubernetes/kubernetes/pull/103257#discussion_r682667876
https://github.com/kubernetes/kubernetes/pull/103257#discussion_r682679708

Split test into #104139 #104140

wzshiming commented:
@derekwaynecarr 

Hi, could you have time to take a look

wzshiming commented:
@derekwaynecarr HI, Could you have time to take a look?
This PR does fix the backoff strategy for container restarts. How can I forward it?

wzshiming commented:
@ehashman Hi

Could you have time to look again? This is indeed a bug in the restart strategy and needs to be fixed.

wzshiming commented:
friendly ping @derekwaynecarr

wzshiming commented:
/retest

wzshiming commented:
/reopen

wzshiming committed:
Add e2e for container restart backoff
